### DISCUSSION TITLE
jupyter kernel crashed when using `quit()` or `exit()`

### DISCUSSION THREAD

**Discussion by meetAndEgg** [microsoft/vscode-jupyter#11121] [Questions and Answers]
I found `quit()` or `exit()` can kill the jupyter kernel.  
Here is my code:
```python
import threading
import time
def work1():
    for i in range(10):
        print('work1')
        time.sleep(0.5)

if __name__ == '__main__':
    thread_obj = threading.Thread(target = work1)
    thread_obj.setDaemon(True)
    thread_obj.start()
    time.sleep(2)
    print('main process over')
    exit()
```
Here is the error log
```
info 17:14:16.308: Cell 13 executed with state Success
error 17:14:16.757: Disposing session as kernel process died ExitCode: undefined, Reason: d:\Anaconda3\envs\python310\lib\site-packages\traitlets\traitlets.py:2202: FutureWarning: Supporting extra quotes around strings is deprecated in traitlets 5.0. You can use 'hmac-sha256' instead of '"hmac-sha256"' if you require traitlets >=5.
  warn(
d:\Anaconda3\envs\python310\lib\site-packages\traitlets\traitlets.py:2157: FutureWarning: Supporting extra quotes around Bytes is deprecated in traitlets 5.0. Use 'cc66fd5b-e4a4-4c37-9ecc-cc48b182bcf0' instead of 'b"cc66fd5b-e4a4-4c37-9ecc-cc48b182bcf0"'.
  warn(

info 17:14:16.758: kill daemon
error 17:14:16.758: Raw kernel process exited code: undefined
```

**Comment by DonJayamanne** [comment]
> found quit() or exit() can kill the jupyter kernel.

thanks for mentioning this,  but I'd like to understand what you were expecting.
ie what's the expected behavior when you run the above code

**Reply by meetAndEgg** [reply]
I am studying `threading` in jupyter notebook, but the result is different from `py` files, which forces me to write code in a  `py` file. So I want to know that is the kernel crash no problem,  and is there anyway to use `exit()` or `quit()` in `ipynb` with the same result in `py`.
Besides, I found another problem: when the main process in `ipynb` is over, the subprocess may create some wrong results.
Here is my code:
```python
from threading import Thread
import time

class MyThread(Thread):
    def run(self):
        for i in range(5):
            print(f'Running{i}...')
            time.sleep(0.5)
            
if __name__ == '__main__':
    t = MyThread()
    t.start()
```
Here is the output
```
Running0...
Running1... Running1... Running2... Running1... Running2... Running3... Running1... Running2... Running3... Running4...
```
Obviously, the correct output is 
```
Running0... Running1... Running2... Running3...Running4...
```

**Reply by DonJayamanne** [reply]
> that is the kernel crash no problem, and is there anyway to use exit() or quit() in ipynb with the same

this is the expected behavior.
when you run exit , you are actually killing the process that's running the kernel, ie the python process is being killed.
hence this is expected.

as for the issue with the output, please could you file an issue and ill look into that.
